localstack/services/apigateway/provider_asf.py

```diff
@@ -0,0 +1,191 @@
+    apigateway API).
+    """
+
+    router: Router[Handler]
+
+    def __init__(self, router: Router[Handler]):
+        self.router_rules: Dict[str, List[Rule]] = defaultdict(list)
+        self.router = router
+
+    def add_rest_api(self, rest_api: RestApi) -> None:
+        """
+        Adds a route for the given RestApi.
+        :param rest_api: the RestApi to add
+        """
+        # TODO: probably it is better to have a parameterized handler and only one rule, rather than creating a new
+        #  rule for every rest API. the more rules there are, the slower the routing will be. on the other hand,
+        #  regex matching could be just as slow. need to check!
+        api_id = rest_api["id"]
+
+        # add the canonical execute-api handler
+        self.router_rules[api_id].append(
+            self.router.add(
+                "/",
+                host=f"{api_id}.execute-api.<regex('.*'):server>",
+                endpoint=self._restapis_host_handler,
+            )
+        )
+        self.router_rules[api_id].append(
+            self.router.add(
+                "/<path:path>",
+                host=f"{api_id}.execute-api.<regex('.*'):server>",
+                endpoint=self._restapis_host_handler,
+            )
+        )
+
+        # add the localstack-specific _user_request_ handler
+        self.router_rules[api_id].append(
+            self.router.add(
+                f"/restapis/{api_id}/<stage>/_user_request_",
+                endpoint=self._restapis_user_request_handler,
+            )
+        )
+        self.router_rules[api_id].append(
+            self.router.add(
+                f"/restapis/{api_id}/<stage>/_user_request_/<path:path>",
+                endpoint=self._restapis_user_request_handler,
+            )
+        )
+
+    def remove_rest_api(self, rest_api_id: str) -> None:
+        """
+        Removes the given rest api.
+        :param rest_api_id: the rest api to remove
+        """
+        rules = self.router_rules.pop(rest_api_id, [])
+        for rule in rules:
+            self.router.remove_rule(rule)
+
+    def _restapis_handler(self, request: Request, path=None) -> Response:
+        invocation_context = to_invocation_context(request)
+
+        result = invoke_rest_api_from_request(invocation_context)
+        if result is not None:
+            if isinstance(result, LambdaResponse):
+                headers = Headers(dict(result.headers))
+                for k, values in result.multi_value_headers.items():
+                    for value in values:
+                        headers.add(k, value)
+            else:
+                headers = dict(result.headers)
+            return Response(
+                response=result.content,
+                status=result.status_code,
+                headers=headers,
+            )
+
+        raise NotFound()
+
+    def _restapis_user_request_handler(self, request: Request, stage=None, path=None):
```

Comment:
nit:`stage` not being used, I believe this should be added to the invocation context.